### PR TITLE
Make items selectable -> editable via config

### PR DESCRIPTION
**Pull request by p3root**
With my patch you can select a item and have a bindable item.

### CODE REVIEW

File: src/lib/gridster.component.ts [hunk 1]
```diff
@@ -50,7 +57,25 @@ export class GridsterComponent implements OnInit, OnChanges, OnDestroy, Gridster
...
+    return this._selectedItem;
+  }
+  public set selectedItem(v: GridsterItem) {
+    this._selectedItem = v;
+    this.selectedItemChange.emit(v);
+  }
+
+
+  @ViewChildren(forwardRef(() => GridsterItemComponent))
+  gridsterItems: QueryList<GridsterItemComponent>;
```
Comment (tiberiuzuld) [inline src/lib/gridster.component.ts]: why do you need this?

File: src/lib/gridster.component.ts [hunk 2]
```diff
@@ -76,6 +101,23 @@ export class GridsterComponent implements OnInit, OnChanges, OnDestroy, Gridster
     if (this.options.initCallback) {
       this.options.initCallback(this);
     }
+
+    if (this.options.selectable && this.options.selectable.enabled) {
+      this.selectionService.selectItem.subscribe((item: GridsterItemComponentInterface) => {
+        for (const x of this.grid) {
```
Comment (tiberiuzuld) [inline src/lib/gridster.component.ts]: why for loop when you can say:
```javascript
if(this.selectedItem){
	item.item.isSelected = false;	
}
```

File: src/lib/gridsterSelection.service.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import { Injectable, EventEmitter } from '@angular/core';
+import { GridsterItem, GridsterItemComponentInterface } from 'lib';
+
+@Injectable()
+export class GridsterSelectionService {
+    public selectItem = new EventEmitter<GridsterItemComponentInterface>();
```
Comment (tiberiuzuld) [inline src/lib/gridsterSelection.service.ts]: don't really like EventEmitters last time I played with them where creating memory leaks.
I prefer callbacks

File: src/lib/gridster.component.ts (hunk 1 above)
Comment (p3root) [reply src/lib/gridster.component.ts]: was my first approach to get to the components. Dead code. Will be removed!

File: src/lib/gridster.component.ts (hunk 2 above)
Comment (p3root) [reply src/lib/gridster.component.ts]: Good point, will change that.

File: src/lib/gridsterSelection.service.ts (hunk 3 above)
Comment (p3root) [reply src/lib/gridsterSelection.service.ts]: Agree with you, depends on the case how you use them. I will add an unsubscribe in the ngOnDestroy function.

File: src/lib/gridsterDraggable.service.ts [hunk 4]
```diff
@@ -243,9 +245,25 @@ export class GridsterDraggable {
       this.mousedown();
       this.touchstart();
     }
+    if (!enableDrag) {
```
Comment (tiberiuzuld) [inline src/lib/gridsterDraggable.service.ts]: why this here ?
you are already in dragMove means you are already making a drag. ...

File: src/lib/gridsterDraggable.service.ts [hunk 5]
```diff
@@ -243,9 +245,25 @@ export class GridsterDraggable {
...
+    }
+  }
+
+  onMouseDown() {
+    if (this.gridsterItem.item.isSelected) {
+      this.selection.unSelectItem.emit(this.gridsterItem);
+    } else {
+      this.selection.selectItem.emit(this.gridsterItem);
+    }
+    this.gridster.updateGrid();
```
Comment (tiberiuzuld) [inline src/lib/gridsterDraggable.service.ts]: why need to update grid on selection?
This method only updates the lines grid from the background.
Yeah name is not the best for this method.

**Comment by tiberiuzuld** [comment]
Hi @p3root ,
Have some overall questions:
1. Why this needs to be in the library, this can be implemented outside of the library easily?
2. The library is implemented with callbacks not EventEmiters. 
Last time I used those, where creating memory leaks ...
You can test it with the destroy gridster button top right.. and check profiler.
3. You inserted in gridster.component more code than needed, prefer that code to be separate in the selection service maybe.

Thanks

File: src/lib/gridster.component.ts [hunk 6]
```diff
@@ -76,6 +101,23 @@ export class GridsterComponent implements OnInit, OnChanges, OnDestroy, Gridster
...
+          x.item.isSelected = false;
+          x.isSelected = false;
+        }
+        item.item.isSelected = true;
```
Comment (tiberiuzuld) [inline src/lib/gridster.component.ts]: `item.item` object is the object the user gives to the grid.
Don't think is a good idea to pollute that object with grid properties
Use `$item`

**Comment by p3root** [comment]
Hey!

1) I think it can be usefull for other users also. Therefore would it be nice to be in the lib.
2) I have removed the callbacks.
3) I have changed that.

**Comment by p3root** [comment]
I have implemented it by my own. Maybe you bring such a feature by yourself sometimes.

**Comment by tiberiuzuld** [comment]
Added a ticket https://github.com/tiberiuzuld/angular-gridster2/issues/267
Probably after 5.0 release
Thank you.

**Pull request closed without merging**